packages/slate-editor/src/extensions/galleries/components/Gallery/Gallery.tsx

```diff
@@ -1,69 +1,274 @@
     });
 
+    const activeTile: Tile<UploadcareImage> | undefined = calculatedLayout.flat()[activeIndex];
+
+    function handleImagesChange(images: GalleryImage[]) {
+        // Strip out the `id` property before passing the data along
+        onImagesChange(images.map((image) => ({ caption: image.caption, file: image.file })));
+    }
+
+    function handleDragStart({ active }: DragStartEvent) {
+        if (!active) {
+            return;
+        }
+
+        setActiveId(active.id);
+    }
+
+    function handleDragEnd({ over }: DragEndEvent) {
+        if (over) {
+            const overIndex = images.findIndex((image) => image.id === over.id);
+            if (activeIndex !== overIndex) {
+                const newImages = arrayMove(images, activeIndex, overIndex);
+                handleImagesChange(newImages);
+            }
+        }
+
+        setActiveId(null);
+    }
+
+    function handleDragCancel() {
+        setActiveId(null);
+    }
+
+    function handleCaptionChange(id: UniqueIdentifier, caption: string) {
+        const newImages = images.map((image) => {
+            if (image.id === id) {
+                return {
+                    ...image,
+                    caption,
+                };
+            }
+
+            return image;
+        });
+
+        handleImagesChange(newImages);
+    }
+
+    async function handleCrop(id: UniqueIdentifier) {
+        const image = images.find((image) => image.id === id);
+        if (!image) {
+            return;
+        }
+
+        const uploadcareImage = UploadcareImage.createFromPrezlyStoragePayload(image.file);
+
+        async function crop() {
+            const filePromises = await UploadcareEditor.upload(editor, {
+                files: [uploadcareImage],
+                crop: true,
+                imagesOnly: true,
+                tabs: [],
+            });
+
+            if (!filePromises) {
+                return undefined;
+            }
+
+            return await awaitUploads(filePromises);
+        }
+
+        const result = await crop();
+        if (!result) {
+            return;
+        }
+
+        const [croppedImage] = result.successfulUploads.map((fileInfo) => {
+            const image = UploadcareImage.createFromUploadcareWidgetPayload(fileInfo);
+            return image.toPrezlyStoragePayload();
+        });
+
+        const newImages = images.map((image) => {
+            if (image.id === id) {
+                return { ...image, file: croppedImage };
+            }
+
+            return image;
+        });
+
+        handleImagesChange(newImages);
+    }
+
+    function handleDelete(id: UniqueIdentifier) {
+        const newImages = images.filter((image) => image.id !== id);
+        handleImagesChange(newImages);
+    }
+
+    const getIndex = useCallback(
+        (id: UniqueIdentifier) => {
+            return images.findIndex((image) => image.id === id);
+        },
+        [images],
+    );
+
+    let tileIndex = 0;
```

Comment:
I'm using this variable to keep track of the current index in the `images` since `calculatedLayout` is using nested arrays (this was the simplest approach pretty much).